rest/blip_client_test.go

```diff
@@ -0,0 +1,205 @@
+package rest
+
+import (
+	"bytes"
+	"encoding/json"
+	"log"
+	"sync"
+	"time"
+
+	blip "github.com/couchbase/go-blip"
+	fleecedelta "github.com/couchbaselabs/go-fleecedelta"
+	"github.com/google/uuid"
+)
+
+type BlipTesterClient struct {
+	Deltas    bool               // Support deltas on the client side
+	Requests  chan *blip.Message // Tests can consume this channel to inspect requests sent to the client
+	Responses chan *blip.Message // Tests can consume this channel to inspect responses sent from the client
+
+	bt       *BlipTester
+	id       string                       // Generated UUID on creation
+	docsLock sync.RWMutex                 // lock for docs map
+	docs     map[string]map[string][]byte // Map of docID to rev ID to bytes
+}
+
+// NewBlipTesterClient returns a client which emulates the behaviour of a CBL client over BLIP.
+func NewBlipTesterClient(bt *BlipTester) (client *BlipTesterClient, err error) {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	btc := BlipTesterClient{
+		id:        id.String(),
+		bt:        bt,
+		docs:      make(map[string]map[string][]byte, 0),
+		Requests:  make(chan *blip.Message, 1024),
+		Responses: make(chan *blip.Message, 1024),
+	}
+
+	bt.blipContext.HandlerForProfile[messageChanges] = func(request *blip.Message) {
+		btc.Requests <- request
+		log.Printf("client got changes request: %v", request)
+		body, err := request.Body()
+		if err != nil {
+			panic(err)
+		}
+
+		wantedRevs := []interface{}{}
+
+		if string(body) != "null" {
+			var changesReqs [][]interface{}
+			err = json.Unmarshal(body, &changesReqs)
+			if err != nil {
+				panic(err)
+			}
+
+			wantedRevs = make([]interface{}, len(changesReqs))
+			// changesReqs == [[sequence, docID, revID, {deleted}, {size (bytes)}], ...]
+			for i, changesReq := range changesReqs {
+				docID := changesReq[1].(string)
+				btc.docsLock.RLock()
+				if revs, haveDoc := btc.docs[docID]; haveDoc {
+					revList := make([]string, 0, len(revs))
+					for revID := range revs {
+						revList = append(revList, revID)
+					}
+					wantedRevs[i] = revList // send back all revs we have for SG to determine the ancestor
```

Comment:
Stick a TODO: in here to limit the number of wantedRevs (knownRevs) to 20, as I think that's the CBL behaviour.